okhttp/src/test/java/okhttp3/internal/io/FileSystemTest.kt

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright (C) 2020 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okhttp3.internal.io
+
+import java.io.File
+import java.io.IOException
+import okhttp3.TestUtil
+import okio.buffer
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.Rule
+import org.junit.Test
+import org.junit.rules.TemporaryFolder
+import org.junit.runner.RunWith
+import org.junit.runners.Parameterized
+import org.junit.runners.Parameterized.Parameter
+import org.junit.runners.Parameterized.Parameters
+
+/**
+ * Test that our file system abstraction is consistent and sufficient for OkHttp's needs. We're
+ * particularly interested in what happens when open files are moved or deleted on Windows.
+ */
+@RunWith(Parameterized::class)
+class FileSystemTest {
+  @Rule @JvmField val temporaryFolder = TemporaryFolder()
+
+  companion object {
+    @Parameters(name = "{0}") @JvmStatic
+    fun parameters(): Collection<Array<Any>> = listOf(
+        arrayOf(FileSystem.SYSTEM, TestUtil.windows),
+        arrayOf(InMemoryFileSystem(), false),
+        arrayOf(WindowsFileSystem(FileSystem.SYSTEM), true),
+        arrayOf(WindowsFileSystem(InMemoryFileSystem()), true)
+    )
+  }
+
+  @Parameter(0) @JvmField
+  var fileSystem: FileSystem = FileSystem.SYSTEM
+
+  @Parameter(1) @JvmField
+  var windows: Boolean = false
```

Comment:
You can constructor-inject these

```kotlin
class FileSystemTest(val fs: FileSystem, val windows: Boolean)
```

No defaults or lateinit needed.

Reply:
Ooooh how did I not know about that syntax? Much nicer. Fixed.